### PR TITLE
feat: update role only by an owner

### PR DESCRIPTION
**Pull request by sshanzel**
- A new permission for updating source member roles
- A mutation to update a member role

### CODE REVIEW

File: src/schema/sources.ts [hunk 1]
```diff
@@ -312,6 +318,26 @@ export const typeDefs = /* GraphQL */ `
       image: Upload
     ): Source! @auth
 
+    """
+    Set the source member's current role
+    """
+    updateMemberRole(
```
Comment (rebelchris) [inline src/schema/sources.ts:324]: Maybe we should name it `changeMemberRole` as it does both update and demote?

File: src/schema/sources.ts [hunk 2]
```diff
@@ -312,6 +318,26 @@ export const typeDefs = /* GraphQL */ `
...
+      """
+      sourceId: ID!
+
+      """
+      Member to update
+      """
+      memberId: ID!
+
+      """
+      Role to update the user to
```
Comment (rebelchris) [inline src/schema/sources.ts:336]: Also would change this if we do the above

File: src/schema/sources.ts [hunk 3]
```diff
@@ -846,6 +872,27 @@ export const resolvers: IResolvers<any, Context> = {
       });
       return { _: true };
     },
+    updateMemberRole: async (
```
Comment (rebelchris) [inline src/schema/sources.ts:875]: And this :)

**Review by rebelchris — CHANGES_REQUESTED** [review]
There is an edge-case you haven't covered now.
It's owner + moderator can actually remove a member (blocked role)

Would you mind adding that edge-case and some test cases for it?

File: src/schema/sources.ts (hunk 1 above)
Comment (sshanzel) [reply src/schema/sources.ts:324]: The update is applicable to demote and promote I believe. I mean, it just simply means change.

**Comment by rebelchris** [comment]
@sshanzel There is no different mutation, deleting a member is simply assigning the `Blocked` role :)
that was my comment on slack when you took the task.

**Comment by sshanzel** [comment]
> @sshanzel There is no different mutation, deleting a member is simply assigning the Blocked role :)

Oh, you are totally right! Let me update this to include "removal" of members.

**Comment by sshanzel** [comment]
@rebelchris let me know what you think! 🚢

File: __tests__/__snapshots__/sources.ts.snap [hunk 4]
```diff
@@ -213,8 +213,8 @@ Object {
         "post_delete",
         "member_remove",
         "edit",
+        "member_role_update",
         "post_limit",
-        "moderator_add",
         "moderator_remove",
```
Comment (rebelchris) [inline __tests__/__snapshots__/sources.ts.snap:218]: Is this still needed seperate?

File: src/schema/sources.ts [hunk 5]
```diff
@@ -865,6 +891,36 @@ export const resolvers: IResolvers<any, Context> = {
...
+      _,
+      { sourceId, memberId, role }: UpdateMemberRoleArgs,
+      ctx,
+    ): Promise<GQLEmptyResponse> => {
+      if (role === SourceMemberRoles.Blocked) {
```
Comment (rebelchris) [inline src/schema/sources.ts:899]: We also have to think about "unblocking" basically it would be when assigning the "Member" role again, so maybe use `.includes`?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Just one more thing for unblocking members which we can tackle here.
Please also add one test case for moderators unblocking a member 🙏

File: __tests__/__snapshots__/sources.ts.snap (hunk 4 above)
Comment (sshanzel) [reply __tests__/__snapshots__/sources.ts.snap:218]: Oh, right. Missed deleting this one. Thanks!

**Review by rebelchris — APPROVED** [review]
Nice work! 🥳

**Review by idoshamun — APPROVED** [review]
Perfect!

**Comment by sshanzel** [comment]
I checked the design and an owner seems to be allowed to block another owner. Fix should be in this commit: https://github.com/dailydotdev/daily-api/pull/1222/commits/e5cafa7b58c9b437df1e17f53eb30bb2c3c6344a
